src/Cardano/Wallet/CoinSelection/Fee.hs

```diff
@@ -0,0 +1,351 @@
+import Cardano.Wallet.CoinSelection.Random
+    ( distance, pickRandom )
+import Cardano.Wallet.Fee
+    ( Fee (..), FeeError (..), FeeOptions (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), TxIn, TxOut (..), UTxO (..), invariant, isValidCoin )
+import Control.Monad.Trans.Class
+    ( lift )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), throwE )
+import Control.Monad.Trans.Maybe
+    ( MaybeT (..), runMaybeT )
+import Crypto.Random.Types
+    ( MonadRandom )
+import Data.Bifunctor
+    ( bimap )
+
+import qualified Data.List as L
+
+-- | Given the coin selection result from a policy run, adjust the outputs
+-- for fees, potentially returning additional inputs that we need to cover
+-- all fees.
+-- We lose the relationship between the transaction outputs and their
+-- corresponding inputs/change outputs here. This is a decision we
+-- may wish to revisit later. For now however note that since
+--
+--      (a) coin selection tries to establish a particular ratio
+--          between payment outputs and change outputs (currently it
+--          aims for an average of 1:1)
+--
+--      (b) coin selection currently only generates a single change
+--          output per payment output, distributing the fee
+--          proportionally across all change outputs is roughly
+--          equivalent to distributing it proportionally over the
+--          payment outputs (roughly, not exactly, because the 1:1
+--          proportion is best effort only, and may in some cases be
+--          wildly different).
+--
+-- Note that for (a) we don't need the ratio to be 1:1, the above
+-- reasoning will remain true for any proportion 1:n. For (b) however,
+-- if coin selection starts creating multiple outputs, and this number
+-- may vary, then losing the connection between outputs and change
+-- outputs will mean that that some outputs may pay a larger
+-- percentage of the fee (depending on how many change outputs the
+-- algorithm happened to choose).
+adjustForFees
+    :: forall m. MonadRandom m
+    => FeeOptions
+    -> UTxO
+    -> CoinSelection
+    -> ExceptT FeeError m CoinSelection
+adjustForFees opt utxo coinSel = do
+    coinSel'@(CoinSelection inps' outs' chgs') <-
+        senderPaysFee opt utxo coinSel
+    let estimatedFee = getFee $ feeUpperBound opt coinSel
+    let actualFee = getFee $ computeFee coinSel'
+
+    -- We enforce the following invariant:
+    --
+    --   estimatedFee < actualFee < 2 * estimatedFee
+    --
+    -- This coefficient (2*...) is mostly taken out of nowhere, but if anything
+    -- go beyond that upper bound, we would know that our algorithm for fee
+    -- reconciliation below is messed up.
+    -- Similarly, the algorithm tries to take money from inputs until it reaches
+    -- the goal fixed by 'estimatedFee'. So, the actualFee just can't be lower
+    -- than the goal.
+    if (   actualFee < estimatedFee
+        || (coinToDouble actualFee) / 2.0 > (coinToDouble estimatedFee)) then
```

Comment:
same remark here about `Double` ... We know the underlying type, so let's favor this one.